keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,740 @@
+
+These properties will be specified in the BucketRequest and follow the same pattern of events as Bucket creation. i.e. Bucket API object will be updated to reflect the properties set in BucketRequest, and then a controller will pick-up these changes and call the appropriate APIs to reflect them in the backend.
+
+# COSI API Reference
+
+## Bucket 
+
+Resource to represent a Bucket in OSP. Buckets are cluster scoped.
+
+```yaml
+Bucket {
+  TypeMeta
+  ObjectMeta
+  
+  Spec BucketSpec {
+    // Provisioner is the name of driver associated with this bucket
+    Provisioner string 
+
+    // DeletionPolicy is used to specify how COSI should handle deletion of this 
+    // bucket. There are 3 possible values:
+    //  - Retain: Indicates that the bucket should not be deleted from the OSP
+    //  - Delete: Indicates that the bucket should be deleted from the OSP
+    //        once all the workloads accessing this bucket are done
+    //  - ForceDelete: Indicates that the bucket and its contents should be 
+    //        deleted from the OSP even if workloads are currently accessing 
+    //        this bucket
+    DeletionPolicy DeletionPolicy 
+    
+    // Name of the BucketClass specified in the BucketRequest
+    // +optional
+    BucketClassName  string 
+    
+    // Name of the BucketRequest that resulted in the creation of this Bucket
+    // +optional
+    BucketRequest corev1.ObjectReference
+    
+    // Protocol is the data API this bucket is expected to adhere to.
+    // The possible values for protocol are:
+    // -  S3: Indicates Amazon S3 protocol
+    // -  Azure: Indicates Microsoft Azure BlobStore protocol
+    // -  GCS: Indicates Google Cloud Storage protocol
+    Protocol Protocol
+    
+    // Parameters is an opaque map for passing in configuration to a driver
+    // for creating the bucket
+    // +optional
+    Parameters map[string]string 
+    
+    // BucketID is the unique id of the bucket in the OSP. This field should be
+    // used to specify a bucket that has been created outside of COSI. 
+    // +optional
+    BucketID string 
+  }
+  
+  Status BucketStatus {
+    // Possible Conditions:
+    // BucketReady indicates that the bucket is ready for consumption
+    // by workloads
+    Conditions []metav1.Condition
+    
+    // BucketID is the unique id for the bucket in the OSP. This field is set by 
+    // the COSI sidecar after a call to create bucket has succeeded.
+    BucketID string
+  }
 }
 ```
 
-## ProvisonerCreateBucket
+## BucketRequest
 
-This call is made to create the bucket in the backend. If a bucket that matches both name and parameters already exists, then OK (success) must be returned. If a bucket by same name, but different parameters is provided, then the appropriate error code `ALREADY_EXISTS` must be returned by the provisioner. The call to *ProvisonerCreateBucket* MUST be idempotent.
+A request to create a Bucket in OSP. BucketRequest is a namespaced resource
```

Comment:
> If two namespaces create a the same BucketRequest where would this state be stored
This isn't a supported/valid state of operations. We don't have a sentinel to prevent this from happening as of now, however, it is easy to add one. 

> have you considered keeping the state in a tag in the bucket or even an object on it? 
Yes. As mentioned above, it is not needed. Moreover, using the data path is something we are generally apprehensive about.